In [2]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
#import folium # map rendering library

print('Libraries imported.')

Libraries imported.


I have read the table with "pd.read_html" from Wikipedia

In [3]:
import pandas as pd

table = pd.read_html("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M")

print(table[0])

            0                 1  \
0    Postcode           Borough   
1         M1A      Not assigned   
2         M2A      Not assigned   
3         M3A        North York   
4         M4A        North York   
5         M5A  Downtown Toronto   
6         M5A  Downtown Toronto   
7         M6A        North York   
8         M6A        North York   
9         M7A      Queen's Park   
10        M8A      Not assigned   
11        M9A         Etobicoke   
12        M1B       Scarborough   
13        M1B       Scarborough   
14        M2B      Not assigned   
15        M3B        North York   
16        M4B         East York   
17        M4B         East York   
18        M5B  Downtown Toronto   
19        M5B  Downtown Toronto   
20        M6B        North York   
21        M7B      Not assigned   
22        M8B      Not assigned   
23        M9B         Etobicoke   
24        M9B         Etobicoke   
25        M9B         Etobicoke   
26        M9B         Etobicoke   
27        M9B       

Converted it to a dataframe

In [4]:
table1=pd.DataFrame(table[0])


I have changed column names

In [5]:
table2=table1.rename(columns=table1.iloc[0]).drop(table1.index[0])
table2

Postcode           Borough  \
1        M1A      Not assigned   
2        M2A      Not assigned   
3        M3A        North York   
4        M4A        North York   
5        M5A  Downtown Toronto   
6        M5A  Downtown Toronto   
7        M6A        North York   
8        M6A        North York   
9        M7A      Queen's Park   
10       M8A      Not assigned   
11       M9A         Etobicoke   
12       M1B       Scarborough   
13       M1B       Scarborough   
14       M2B      Not assigned   
15       M3B        North York   
16       M4B         East York   
17       M4B         East York   
18       M5B  Downtown Toronto   
19       M5B  Downtown Toronto   
20       M6B        North York   
21       M7B      Not assigned   
22       M8B      Not assigned   
23       M9B         Etobicoke   
24       M9B         Etobicoke   
25       M9B         Etobicoke   
26       M9B         Etobicoke   
27       M9B         Etobicoke   
28       M1C       Scarborough   
29       M1C       Scarborough   
30       M1C       Scarborough   
31       M2C      Not assigned   
32       M3C        North York   
33       M3C        North York   
34       M4C         East York   
35       M5C  Downtown Toronto   
36       M6C              York   
37       M7C      Not assigned   
38       M8C      Not assigned   
39       M9C         Etobicoke   
40       M9C         Etobicoke   
41       M9C         Etobicoke   
42       M9C         Etobicoke   
43       M1E       Scarborough   
44       M1E       Scarborough   
45       M1E       Scarborough   
46       M2E      Not assigned   
47       M3E      Not assigned   
48       M4E      East Toronto   
49       M5E  Downtown Toronto   
50       M6E              York   
51       M7E      Not assigned   
52       M8E      Not assigned   
53       M9E      Not assigned   
54       M1G       Scarborough   
55       M2G      Not assigned   
56       M3G      Not assigned   
57       M4G         East York   
58       M5G  Downtown Toronto   
59       M6G  Downtown Toronto   
60       M7G      Not assigned   
61       M8G      Not assigned   
62       M9G      Not assigned   
63       M1H       Scarborough   
64       M2H        North York   
65       M3H        North York   
66       M3H        North York   
67       M3H        North York   
68       M4H         East York   
69       M5H  Downtown Toronto   
70       M5H  Downtown Toronto   
71       M5H  Downtown Toronto   
72       M6H      West Toronto   
73       M6H      West Toronto   
74       M7H      Not assigned   
75       M8H      Not assigned   
76       M9H      Not assigned   
77       M1J       Scarborough   
78       M2J        North York   
79       M2J        North York   
80       M2J        North York   
81       M3J        North York   
82       M3J        North York   
83       M4J         East York   
84       M5J  Downtown Toronto   
85       M5J  Downtown Toronto   
86       M5J  Downtown Toronto   
87       M6J      West Toronto   
88       M6J      West Toronto   
89       M7J      Not assigned   
90       M8J      Not assigned   
91       M9J      Not assigned   
92       M1K       Scarborough   
93       M1K       Scarborough   
94       M1K       Scarborough   
95       M2K        North York   
96       M3K        North York   
97       M3K        North York   
98       M4K      East Toronto   
99       M4K      East Toronto   
100      M5K  Downtown Toronto   
101      M5K  Downtown Toronto   
102      M6K      West Toronto   
103      M6K      West Toronto   
104      M6K      West Toronto   
105      M7K      Not assigned   
106      M8K      Not assigned   
107      M9K      Not assigned   
108      M1L       Scarborough   
109      M1L       Scarborough   
110      M1L       Scarborough   
111      M2L        North York   
112      M2L        North York   
113      M3L        North York   
114      M4L      East Toronto   
115      M4L      East Toronto   
116      M5L  Downtown Toronto   
117      M5L  Downtown Tor

In [6]:
table3=table2[table2.Borough != 'Not assigned']
table3.reset_index(drop=True, inplace=True)
table3

Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
3        M5A  Downtown Toronto   
4        M6A        North York   
5        M6A        North York   
6        M7A      Queen's Park   
7        M9A         Etobicoke   
8        M1B       Scarborough   
9        M1B       Scarborough   
10       M3B        North York   
11       M4B         East York   
12       M4B         East York   
13       M5B  Downtown Toronto   
14       M5B  Downtown Toronto   
15       M6B        North York   
16       M9B         Etobicoke   
17       M9B         Etobicoke   
18       M9B         Etobicoke   
19       M9B         Etobicoke   
20       M9B         Etobicoke   
21       M1C       Scarborough   
22       M1C       Scarborough   
23       M1C       Scarborough   
24       M3C        North York   
25       M3C        North York   
26       M4C         East York   
27       M5C  Downtown Toronto   
28       M6C              York   
29       M9C         Etobicoke   
30       M9C         Etobicoke   
31       M9C         Etobicoke   
32       M9C         Etobicoke   
33       M1E       Scarborough   
34       M1E       Scarborough   
35       M1E       Scarborough   
36       M4E      East Toronto   
37       M5E  Downtown Toronto   
38       M6E              York   
39       M1G       Scarborough   
40       M4G         East York   
41       M5G  Downtown Toronto   
42       M6G  Downtown Toronto   
43       M1H       Scarborough   
44       M2H        North York   
45       M3H        North York   
46       M3H        North York   
47       M3H        North York   
48       M4H         East York   
49       M5H  Downtown Toronto   
50       M5H  Downtown Toronto   
51       M5H  Downtown Toronto   
52       M6H      West Toronto   
53       M6H      West Toronto   
54       M1J       Scarborough   
55       M2J        North York   
56       M2J        North York   
57       M2J        North York   
58       M3J        North York   
59       M3J        North York   
60       M4J         East York   
61       M5J  Downtown Toronto   
62       M5J  Downtown Toronto   
63       M5J  Downtown Toronto   
64       M6J      West Toronto   
65       M6J      West Toronto   
66       M1K       Scarborough   
67       M1K       Scarborough   
68       M1K       Scarborough   
69       M2K        North York   
70       M3K        North York   
71       M3K        North York   
72       M4K      East Toronto   
73       M4K      East Toronto   
74       M5K  Downtown Toronto   
75       M5K  Downtown Toronto   
76       M6K      West Toronto   
77       M6K      West Toronto   
78       M6K      West Toronto   
79       M1L       Scarborough   
80       M1L       Scarborough   
81       M1L       Scarborough   
82       M2L        North York   
83       M2L        North York   
84       M3L        North York   
85       M4L      East Toronto   
86       M4L      East Toronto   
87       M5L  Downtown Toronto   
88       M5L  Downtown Toronto   
89       M6L        North York   
90       M6L        North York   
91       M6L        North York   
92       M9L        North York   
93       M1M       Scarborough   
94       M1M       Scarborough   
95       M1M       Scarborough   
96       M2M        North York   
97       M2M        North York   
98       M3M        North York   
99       M4M      East Toronto   
100      M5M        North York   
101      M5M        North York   
102      M6M              York   
103      M6M              York   
104      M6M              York   
105      M6M              York   
106      M9M        North York   
107      M9M        North York   
108      M1N       Scarborough   
109      M1N       Scarborough   
110      M2N        North York   
111      M3N        North York   
112      M4N   Central Toronto   
113      M5N   Central Toronto   
114      M6N              York   
115      M6N              York   
116      M9N              

In [7]:
table4=table3.replace(['Not assigned'], ["Queen's Park"])
table4

Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
3        M5A  Downtown Toronto   
4        M6A        North York   
5        M6A        North York   
6        M7A      Queen's Park   
7        M9A         Etobicoke   
8        M1B       Scarborough   
9        M1B       Scarborough   
10       M3B        North York   
11       M4B         East York   
12       M4B         East York   
13       M5B  Downtown Toronto   
14       M5B  Downtown Toronto   
15       M6B        North York   
16       M9B         Etobicoke   
17       M9B         Etobicoke   
18       M9B         Etobicoke   
19       M9B         Etobicoke   
20       M9B         Etobicoke   
21       M1C       Scarborough   
22       M1C       Scarborough   
23       M1C       Scarborough   
24       M3C        North York   
25       M3C        North York   
26       M4C         East York   
27       M5C  Downtown Toronto   
28       M6C              York   
29       M9C         Etobicoke   
30       M9C         Etobicoke   
31       M9C         Etobicoke   
32       M9C         Etobicoke   
33       M1E       Scarborough   
34       M1E       Scarborough   
35       M1E       Scarborough   
36       M4E      East Toronto   
37       M5E  Downtown Toronto   
38       M6E              York   
39       M1G       Scarborough   
40       M4G         East York   
41       M5G  Downtown Toronto   
42       M6G  Downtown Toronto   
43       M1H       Scarborough   
44       M2H        North York   
45       M3H        North York   
46       M3H        North York   
47       M3H        North York   
48       M4H         East York   
49       M5H  Downtown Toronto   
50       M5H  Downtown Toronto   
51       M5H  Downtown Toronto   
52       M6H      West Toronto   
53       M6H      West Toronto   
54       M1J       Scarborough   
55       M2J        North York   
56       M2J        North York   
57       M2J        North York   
58       M3J        North York   
59       M3J        North York   
60       M4J         East York   
61       M5J  Downtown Toronto   
62       M5J  Downtown Toronto   
63       M5J  Downtown Toronto   
64       M6J      West Toronto   
65       M6J      West Toronto   
66       M1K       Scarborough   
67       M1K       Scarborough   
68       M1K       Scarborough   
69       M2K        North York   
70       M3K        North York   
71       M3K        North York   
72       M4K      East Toronto   
73       M4K      East Toronto   
74       M5K  Downtown Toronto   
75       M5K  Downtown Toronto   
76       M6K      West Toronto   
77       M6K      West Toronto   
78       M6K      West Toronto   
79       M1L       Scarborough   
80       M1L       Scarborough   
81       M1L       Scarborough   
82       M2L        North York   
83       M2L        North York   
84       M3L        North York   
85       M4L      East Toronto   
86       M4L      East Toronto   
87       M5L  Downtown Toronto   
88       M5L  Downtown Toronto   
89       M6L        North York   
90       M6L        North York   
91       M6L        North York   
92       M9L        North York   
93       M1M       Scarborough   
94       M1M       Scarborough   
95       M1M       Scarborough   
96       M2M        North York   
97       M2M        North York   
98       M3M        North York   
99       M4M      East Toronto   
100      M5M        North York   
101      M5M        North York   
102      M6M              York   
103      M6M              York   
104      M6M              York   
105      M6M              York   
106      M9M        North York   
107      M9M        North York   
108      M1N       Scarborough   
109      M1N       Scarborough   
110      M2N        North York   
111      M3N        North York   
112      M4N   Central Toronto   
113      M5N   Central Toronto   
114      M6N              York   
115      M6N              York   
116      M9N              

I have completed all data cleaning works

In [18]:
import numpy as np  
import pandas as pd 
table4['Neighbourhood']=table4['Neighbourhood'].astype(str)
result=table4.groupby(['Postcode','Borough'], sort=False).agg(','.join).reset_index()
result

Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
3        M6A        North York   
4        M7A      Queen's Park   
5        M9A         Etobicoke   
6        M1B       Scarborough   
7        M3B        North York   
8        M4B         East York   
9        M5B  Downtown Toronto   
10       M6B        North York   
11       M9B         Etobicoke   
12       M1C       Scarborough   
13       M3C        North York   
14       M4C         East York   
15       M5C  Downtown Toronto   
16       M6C              York   
17       M9C         Etobicoke   
18       M1E       Scarborough   
19       M4E      East Toronto   
20       M5E  Downtown Toronto   
21       M6E              York   
22       M1G       Scarborough   
23       M4G         East York   
24       M5G  Downtown Toronto   
25       M6G  Downtown Toronto   
26       M1H       Scarborough   
27       M2H        North York   
28       M3H        North York   
29       M4H         East York   
30       M5H  Downtown Toronto   
31       M6H      West Toronto   
32       M1J       Scarborough   
33       M2J        North York   
34       M3J        North York   
35       M4J         East York   
36       M5J  Downtown Toronto   
37       M6J      West Toronto   
38       M1K       Scarborough   
39       M2K        North York   
40       M3K        North York   
41       M4K      East Toronto   
42       M5K  Downtown Toronto   
43       M6K      West Toronto   
44       M1L       Scarborough   
45       M2L        North York   
46       M3L        North York   
47       M4L      East Toronto   
48       M5L  Downtown Toronto   
49       M6L        North York   
50       M9L        North York   
51       M1M       Scarborough   
52       M2M        North York   
53       M3M        North York   
54       M4M      East Toronto   
55       M5M        North York   
56       M6M              York   
57       M9M        North York   
58       M1N       Scarborough   
59       M2N        North York   
60       M3N        North York   
61       M4N   Central Toronto   
62       M5N   Central Toronto   
63       M6N              York   
64       M9N              York   
65       M1P       Scarborough   
66       M2P        North York   
67       M4P   Central Toronto   
68       M5P   Central Toronto   
69       M6P      West Toronto   
70       M9P         Etobicoke   
71       M1R       Scarborough   
72       M2R        North York   
73       M4R   Central Toronto   
74       M5R   Central Toronto   
75       M6R      West Toronto   
76       M7R       Mississauga   
77       M9R         Etobicoke   
78       M1S       Scarborough   
79       M4S   Central Toronto   
80       M5S  Downtown Toronto   
81       M6S      West Toronto   
82       M1T       Scarborough   
83       M4T   Central Toronto   
84       M5T  Downtown Toronto   
85       M1V       Scarborough   
86       M4V   Central Toronto   
87       M5V  Downtown Toronto   
88       M8V         Etobicoke   
89       M9V         Etobicoke   
90       M1W       Scarborough   
91       M4W  Downtown Toronto   
92       M5W  Downtown Toronto   
93       M8W         Etobicoke   
94       M9W         Etobicoke   
95       M1X       Scarborough   
96       M4X  Downtown Toronto   
97       M5X  Downtown Toronto   
98       M8X         Etobicoke   
99       M4Y  Downtown Toronto   
100      M7Y      East Toronto   
101      M8Y         Etobicoke   
102      M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                             Harbourfront,Regent Park  
3                      Lawrence Heights,Lawrence Manor  
4                                         Queen's Park  
5                                     Islington Avenue  
6                                        Rouge,Malvern  
7          

There are 103 rows and 3 columns in my final dataframe

In [19]:
result.shape

(103, 3)